### DISCUSSION TITLE
blog/seeding-the-query-cache

### DISCUSSION THREAD

**Discussion by giscus** [TkDodo/blog-comments#86] [Announcements]
# blog/seeding-the-query-cache

With suspense for data fetching on the horizon, it is now more important than ever to make sure your cache is seeded properly to avoid fetch waterfalls.

https://tkdodo.eu/blog/seeding-the-query-cache

**Comment by Moshyfawn** [comment]
Had to deal with the exact use-case recently :D I would've linked this post as my PR description if it existed at the time lol

**Comment by marcortw** [comment]
I am playing around with the push approach and I somehow can't get the staleTime to be automatically respected. Would you mind explaining what exactly you mean with that? My expectation, through this sentence was somehow, that the staleTime option would also apply automagically to the data added with setQueryData.

```
export function useTeams() {
  const queryClient = useQueryClient();
  return useQuery({
    queryKey: ["teams", "list"],
    staleTime: /* Infinity */ 30 * 1000,
    queryFn: async () => {
      const teams = await fetchAllTeams();
      teams.forEach((team) => {
        // Push approach as described on https://tkdodo.eu/blog/seeding-the-query-cache#push-approach
        queryClient.setQueryData(["teams", "details", team.teamId], team);
      });
      return teams;
    },
  });
}
```
The result for me is that ["teams", "details", team.teamId] is not having the staleTime option set, it seems to be independent of the ["teams", "list"] queryKey. It is seeded as inactive, which seems to be fine, but I don't understand what is now exactly meant as "staleTime is automatically respected" other than the staleTime option of a detail cache entry. Thanks in advance!

**Reply by TkDodo** [reply]
I meant the following: The `["teams", "details", team.teamId]` is created at the time when `useTeams()` resolves. Let's say this happens at 09:00. Now, if you mount a component with `useQuery` that uses this entry 2 minutes later at 09:02, it would still be fresh IF you provide a staleTime of e.g. 5 minutes. If you mount that component at 09:06 instead, it would be considered stale because data is 6 minutes old already.

Compared to the pull approach, where `initialData` is used, if you'd read from the `["teams", "list"]` at 09:06 when the query is created, it would be fresh for 5 minutes _from that point in time_, so until 09:11. That's why you'd need `initialDataUpdatedAt` with that approach.

So `staleTime is automatically respected` means that staleTime will work correctly if you supply a staleTime 😅 . Hope it makes more sense now.

**Reply by marcortw** [reply]
Thanks a lot for the additional explanation, @TkDodo ! I think I understand it now. I'm new to tanstack query and greatly appreciate the extensive docs, your blog posts and community support :+1:

**Comment by EnshaednHiker** [comment]
What about prefetching multiple queries server side? I don't see a `pretetch` equivalent to `useQueries`. Is it still performant to call multiple `prefetch`es server side or do you have a different recommendation?

**Reply by TkDodo** [reply]
you can call `prefetch` multiple times and await them in a `Promise.all`:

```
const promise1 = queryClient.prefetchQuery(...)
const promise2 = queryClient.prefetchQuery(...)

await Promise.all([promise1, promise2]
```

**Comment by williambalea** [comment]
In your pull approach example, you used `useQueryClient` to get the `data` and the `dataUpdatedAt`. Is it okay to use 
```const { data, dataUpdatedAt } = useQuery({queryKey: ['todos', 'list'], ...})``` instead ?

**Reply by TkDodo** [reply]
that will create another subscription to the list query

**Comment by layerok** [comment]
I love yout articles, everytime I find out something new, thank you.

**Comment by liornz** [comment]
As always, a great post. I think it could be useful to mention that the choice between the pull and the push methods depends also on the use case. Not always are both options equally effective. It depends how the components which subscribe to the queries are composed, howcabd when they are rendered and rerendered and when each query runs.

**Comment by shahbazshueb** [comment]
This is amazing. I just wanted to do the same for a long list that I am fetching.

**Comment by heesoopark5473** [comment]
Hi! Love you article. 

I have a question, 

I use `prefetchQuery` when user opens the app to have some setting prepared for usage in the cache. 

`
        await queryClient.prefetchQuery({
          queryKey: ['user-preference', _address],
          queryFn: () => userPreferencesFn(_address),
          cacheTime: Infinity,
        });
`

The cacheTime is Infinity, so the cached objects are alive throughout the usage of the app. 

I also have 

`
export const useUserPreferences = (wallet_address?: string) => {
  const queryKey = ['user-preference', wallet_address];
  return useQuery({
    queryKey,
    queryFn: () => userPreferencesFn(wallet_address),
    enabled: Boolean(wallet_address),
    cacheTime: Infinity,
  });
};
`

Which fetches the latest preferences  in one of the components. 

I was wondering, since I had set the cacheTime to Infinity, do I not need `cacheTime: Infinity` in `useUserPreferences`? 
If I was to not have `cacheTime` option in `useUserPreferences`, would cacheTime still be Infinity?

I am little bit confused because if I do not set cacheTime in useQuery or if I set cacheTime to 50 in useQuery,  it seems app behaves as if cacheTime is Infinity, as set in prefetch. 

But, when i console.log the cacheTime via queryClient.getQueryData, it shows the latest updated cacheTIme. 

Then, I look at the source code, and cacheTime seems take the Max of existing value and newValue, so any new cacheTime smaller then existing cacheTime. would be discarded. 

I am using react-query v4 in react-native.

Thanks for all your contribution!

**Reply by TkDodo** [reply]
The source code is right 👍

**Reply by heesoopark5473** [reply]
Thank you very much for your response!

Now I understand that once cacheTime is set to Infinity, any other cacheTime < Inifnity would be discarded and cacheTime remains as Inifinity.

Was just confused because, { cacheTime } = queryClient.getQueryData would return the latested updated cachTime value, not Infinity. 

But maybe queryClient.getQueryData is using different reference to console.log.

**Reply by TkDodo** [reply]
`queryClient.getQueryData` returns data stored in the cache. It doesn’t return any options, so unless your `queryFn` returns an object with a field called `cacheTime`, I wouldn’t expect this to log anything.

Regardless, what you return from your `queryFn` is totally independent of any config settings.

**Comment by aboqasem** [comment]
Thanks for this great article!

I think other issues with the "Pull" approach is that:

🟡   it cannot be implemented the other way around
🔴   item state may divert from the listing state

---

And the following issues with the "Push" approach may be avoided:

🟡   might create unnecessary cache entries
🔴   pushed data might be garbage collected too early

By adding some checks, simplest one being that whether or not the data already exists, which acts as an update mechanism:

```diff
-       queryClient.setQueryData(['todos', 'detail', todo.id], todo)
+       const itemQueryKey = ['todos', 'detail', todo.id];
+       if (queryClient.getQueryData(itemQueryKey)) queryClient.setQueryData(itemQueryKey, todo)
```

What do you think?

**Reply by TkDodo** [reply]
Your suggestion would only update existing items, which isn’t what “seeding” is about. We’d definitely want to set data for non-existing items, otherwise, we’ll:

- open the list, seed nothing because no detail entry exists and nothing is written
- we navigate to a detail item
- we get a hard loading state because we have no data

This is the flow we’re trying to avoid with seeding the cache.

**Comment by MhykeB** [comment]
Thanks for the article it's very helpful and insightful. I came here while i was reading the tanstack docs to find a solution to an issue I've been experiencing. Hopeful you might be able to help me out. 
 When i prefetch from the server in a server component and i hydrate my query client with the data returned, using a unique query key. It works as expected, the client component sees the data in the cache and doesn't bother refetching. Note that i wrap my server component in a suspense boundary to so that the page shows up more quickly with a loading effect fall back. 
 My issue is that when i navigate away from this page and go back to it, the data is prefetched again from the server and i see a loading ui from the suspense boundary fallback, even though the query is still present in the cache. This is different from when i fetch the data entirely client side, the former data is used while a background fetch happens, the user doesn't get to see a loading UI. 
I guess there's something I'm missing. 
 I'd anticipate your response, thanks.

**Reply by TkDodo** [reply]
React Query doesn’t cache on the server - every request gets a new, empty QueryClient. So _if_ the server component re-evaluates, you will get another fetch.

The best way around this is to make sure the server component doesn’t re-run when you don’t want to. Next has a lot of caching options for this, I don’t know which one you need to set, but I think they also have a `staleTime` now.

Note that this is differently in other frameworks like TanStack Start, where every navigation is a fully client-side navigation after the initial SSR.
